### ISSUE TITLE
Deploying an upgrade

### ISSUE THREAD

**Issue by ignasi35** [lightbend/openshift-deployment-guide#50]
When a service is already running smoothly and anew version of the code is deployed we want new pods to join the existing cluster and do a rolling upgrade.

LBO supported `rp generate-kubernetes-resources ... --join-existing-akka-cluster` which disables Akka's [`form-new-cluster`](https://github.com/akka/akka-management/blob/e6582043b54a0fa8443477a68600d1ed14faa756/cluster-bootstrap/src/main/resources/reference.conf#L21). The Akka `form-new-cluster` documentation [states](https://github.com/akka/akka-management/blob/e6582043b54a0fa8443477a68600d1ed14faa756/cluster-bootstrap/src/main/resources/reference.conf#L17-L18):

> Setting `new-cluster-enabled=off` after an initial cluster has formed is recommended to prevent new clusters forming during a network partition when nodes are redeployed or restarted.

This guide should cover the necessary steps and settings to complete a rolling upgrade of a clustered application safely.

**Comment by TimMoore** [comment]
Related to #20

**Comment by TimMoore** [comment]
Note that the property was renamed from `form-new-cluster` to `new-cluster-enabled` to reduce confusion. Whether it's enabled or disabled, it will attempt to join an existing cluster first, and only form a new one if no seed nodes are found on the node with the lowest address. It isn't required to change this property on redeployment, only recommended as an additional safety mechanism.

**Comment by ignasi35** [comment]
My main doubt is: (the following is my limited understanding)

 * if a new `Deployment` with `new-cluster-enabled=false` is `apply`ed on kubernetes/openshift that'll cause a rolling upgrade. That's what we usually want. But, that action will leave a `Deployment` on the cluster with `new-cluster-enabled` disabled. As a consequence, if/when the Akka cluster crashes and k8s brings up new pods they won't form a cluster. Is that correct?

I think I'm missing something but it seems to me like the value of `new-cluster-enabled` should be determined by kubernetes dynamically depending on the existence of an Akka Cluster.

**Comment by TimMoore** [comment]
> if a new Deployment with new-cluster-enabled=false is applyed on kubernetes/openshift that'll cause a rolling upgrade. That's what we usually want. But, that action will leave a Deployment on the cluster with new-cluster-enabled disabled.

Maybe I don't understand what you mean. Wouldn't the new `Deployment` replace the existing one? You wouldn't want two different deployments for the same service/cluster.

> I think I'm missing something but it seems to me like the value of new-cluster-enabled should be determined by kubernetes dynamically depending on the existence of an Akka Cluster.

That would be nice, but I don't think there is anything exposed from Kubernetes to a container that indicates whether it is part of a new deployment or an update/scale up/replacement. I'm not sure it's even possible in principle without knowledge of Akka Cluster specifics, since the initial cluster could fail to form for some reason and trigger a restart of the containers, and this wouldn't be externally distinguishable from restarting the containers after the cluster forms. What this setting indicates is really the operator's intent. I think only a custom Kubernetes Operator would be able to handle this more automatically.

**Comment by jroper** [comment]
The whole point of the feature is to put the formation of a new cluster into the users hands. Custom operator or not, Kubernetes can't 100% be relied on to tell you if there's an existing cluster, you'll have to talk to the Akka team about this, but as I understand it, in an etcd split brain scenario, new writes are prevented on the minority side, but reads don't fail, so you could have a scenario where there is an Akka cluster whose majority (or whichever nodes Akka SBR will decide to keep) is running on the etcd minority side of the partition, while the majority side will decide to start up a new one, and you'll get two clusters. In these scenarios, there is no difference between an operator making that decision to bootstrap a new cluster, and Akka cluster bootstrap making that decision based on querying the Kubernetes API - both are using the same source of data (Kubernetes API - etcd) for their decisions, so an operator setting the new cluster enabled flag or not won't make any difference to what would have happened if cluster bootstrap were allowed to make the decision by itself.

So the idea behind the feature is if you really want to protect against those scenarios, then allow the user to specifically request the bootstrap of a new cluster. Then, as long as the user makes the right decision, you're guaranteed to not get two clusters. In the above scenario, the first thing a user is going to focus on is the partition itself. Once that is resolved, then I guess k8s is either going to incorporate the remaining cluster on the minority side into its current state, or perhaps kill it, not sure which it does, but either way, the user will then be able to check the status of the app, and if no cluster is running, they can explicitly set the flag to start a new cluster and deploy. The downside, is that there are times when the entire cluster will crash, and the problem could be resolved by k8s restarting it, but because of the flag, it will always require user intervention. That's a trade off that the user has to make, that's always the trade off, you can have consistency or availability in the face of network partitions, not both.

**Comment by ignasi35** [comment]
> Maybe I don't understand what you mean. Wouldn't the new Deployment replace the existing one? You wouldn't want two different deployments for the same service/cluster.

That is exactly what I mean. There will never be two `Deployment`s.

When the _new_ `Deployment` with a `new-cluster-enabled` disabled is applied, the old `Deployment` is gone. If the k8s cluster crashes or the Akka cluster reboots (e.g. [SBR `down-all` strategy](https://developer.lightbend.com/docs/akka-commercial-addons/current/split-brain-resolver.html#down-all)) the Akka cluster will never restart because the existing `Deployment` has the setting `new-cluster-enabled` disabled. A human intervention is then required.

**Comment by TimMoore** [comment]
Yes, that's true. It's a trade-off that you'll need to decide on. It would be good to explain the considerations in the guide.
